src/modules/flow/freegeoip/freegeoip.c

```diff
@@ -0,0 +1,239 @@
+        } \
+    } while (0)
+
+enum fields {
+    FIELD_IP = 1<<0,
+    FIELD_COUNTRY_CODE = 1<<1,
+    FIELD_COUNTRY_NAME = 1<<2,
+    FIELD_REGION_CODE = 1<<3,
+    FIELD_CITY_NAME = 1<<4,
+    FIELD_ZIP_CODE = 1<<5,
+    FIELD_TIMEZONE = 1<<6,
+    FIELD_LATITUDE = 1<<7,
+    FIELD_LONGITUDE = 1<<8,
+    FIELD_ALL = 0x1ff
+};
+
+static bool
+json_token_to_double(const struct sol_json_token *token, double *output)
+{
+    char *str = strndupa(token->start, token->end - token->start);
+    char *endptr;
+    double value;
+
+    errno = 0;
+    value = strtod(str, &endptr);
+    if (errno != 0)
+        return false;
+    if (!endptr || *endptr)
+        return false;
+    *output = value;
+    return true;
+}
+
+static void
+freegeoip_query_finished(void *data, struct sol_http_response *response)
+{
+    struct freegeoip_data *mdata = data;
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    enum fields fields = FIELD_ALL;
+    double latitude = 0.0f;
+    double longitude = 0.0f;
+
+    RESPONSE_CHECK_API(response, mdata);
+
+    if (!response->content.used) {
+        SOL_WRN("Empty response from FreeGeoIP");
+        return;
+    }
+
+    if (response->response_code != 200) {
+        sol_flow_send_error_packet(mdata->node, EINVAL,
+            "FreeGeoIP returned an unknown response code: %d",
+            response->response_code);
+        return;
+    }
+
+#define FIELD(json_field_, flow_field_) \
+    do { \
+        if (fields & FIELD_ ## flow_field_ && sol_json_token_str_eq(&key, \
+                json_field_, strlen(json_field_))) { \
+            fields &= ~(FIELD_ ## flow_field_); \
+            sol_flow_send_string_packet(mdata->node, \
+                SOL_FLOW_NODE_TYPE_LOCATION_FREEGEOIP__OUT__ ## flow_field_, \
+                strndupa(value.start + 1, value.end - value.start - 2)); \
+            continue; \
+        } \
+    } while (0)
+
+    sol_json_scanner_init(&scanner, response->content.data, response->content.used);
+    SOL_JSON_SCANNER_OBJECT_LOOP(&scanner, &token, &key, &value, reason) {
+        FIELD("ip", IP);
+        FIELD("country_name", COUNTRY_NAME);
+        FIELD("country_code", COUNTRY_CODE);
+        FIELD("city", CITY_NAME);
+        FIELD("zip_code", ZIP_CODE);
+        FIELD("time_zone", TIMEZONE);
+
+        if (fields & FIELD_LATITUDE && sol_json_token_str_eq(&key, "latitude", strlen("latitude"))) {
+            if (json_token_to_double(&value, &latitude))
+                fields &= ~FIELD_LATITUDE;
+            continue;
+        }
+
+        if (fields & FIELD_LONGITUDE && sol_json_token_str_eq(&key, "longitude", strlen("longitude"))) {
+            if (json_token_to_double(&value, &longitude))
+                fields &= ~FIELD_LONGITUDE;
+            continue;
+        }
+    }
+
+#undef FIELD
+
+    if (reason != SOL_JSON_LOOP_REASON_OK)
+        return;
+
+    if (!(fields & (FIELD_LATITUDE | FIELD_LONGITUDE))) {
```

Comment:
since you only use 'fields' here and it's used only to send latitude or longitude, what if we initialize them with FP_NAN or even check if both are FP_ZERO?